rfd/0192-oci-join.md

```diff
@@ -0,0 +1,286 @@
+ID, and instance ID respectively.
+- The auth server validates/verifies several properties:
+  - The tenancy ID, compartment ID, and instance ID are all valid Oracle OCIDs.
+  - The tenancy ID, compartment ID, and region match the Teleport provision token. 
+  - The signed challenge matches.
+- The auth server makes a HTTP request to ListCompartments, using the region from
+the JWT (found in the instance OCID) and the signed headers.
+```go
+apiReq, err := http.NewRequest(
+  http.MethodGet,
+  fmt.Sprintf("https://iaas.%s.oraclecloud.com/20160918", region /* sanitized */),
+  nil)
+apiReq.Header = apiReqInfo.Header
+```
+If Oracle accepts it, the node is allowed to join the cluster.
+
+#### Throttling
+
+If the auth server is ever
+[throttled by Oracle](https://docs.oracle.com/en-us/iaas/Content/API/Concepts/usingapi.htm#throttle),
+the TooManyRequests error will be propagated back to the node, which will try
+`RegisterUsingOracleMethod` again with the exponential backoff recommended by
+Oracle (maximum of 60 seconds).
+
+#### Limitations
+
+The Oracle provision tokens will not support nested compartments, i.e. if
+compartment `foo` has a child compartment `bar` and the provision token has
+`parent_compartments: ["foo"]`, this will not allow instances in container `bar` to
+join. This is for simplicity's sake; Teleport would need to make several
+requests to the Oracle Cloud API to walk up the compartment tree from the
+compartment the instance is in, each of which would need to be signed. This
+would require a complicated back-and-forth between the auth server and the
+joining node to get signed requests for each compartment.
+
+### Security
+
+To mitigate SSRF, Teleport will sanitize the claims in the key ID JWT provided by the instance
+to ensure that they are
+[properly formed OCIDs](https://docs.oracle.com/en-us/iaas/Content/General/Concepts/identifiers.htm),
+including verifying that the region in the instance OCID is valid.
+
+On top of the signed challenge, Teleport will verify that the JWT is not expired
+and was not issued in the future (with a leeway of 2 minutes, the same as the
+other JWT-based join methods) to mitigate replay attacks. The Oracle API will
+also verify that the Date header in the signed request is
+[within 5 minutes](https://docs.oracle.com/en-us/iaas/Content/API/Concepts/usingapi.htm#clock)
+of its own clock.
+
+Teleport will not verify signature of the JWT. This is because the needed JWKs will always be behind a
+[non-public API](https://docs.oracle.com/en-us/iaas/Content/APIGateway/Tasks/apigatewayusingjwttokens.htm#identityprovider).
+Instead, Teleport will trust the response from the Oracle Cloud API to know if
+it can trust the signature.
+
+### Proto Specification
+
+Add `RegisterUsingOracleMethod` rpc to the join service:
+
+```proto
+message RegisterUsingOracleMethodRequest { 
+  types.RegisterUsingTokenRequest register_using_token_request = 1;
+  string url = 2;
```

Comment:
It was, but I've realized with your other comment that just sending the region is safer.

Reply:
Does the region need to be supplied by the client? It's my understanding that it will be present in both the configuration of the `token` resource and the JWT. It can be safely extracted and validated given both sources.

Or is there another use case I'm overlooking?